### PR TITLE
Fix the potential access to upsert metadata manager after it is closed

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Even though very unlikely, there is still a code path which can potentially access the upsert metadata after the manager is closed:
- Thread 1 finishes the `_stopped` flag check but not invoking `startOperation()` yet
- Thread 2 stops and closes the metadata manager
- Thread 1 calls `startOperation()` and start accessing the metadata

In this PR, make `startOperation()` return false if the pending operations (reference count) already drops to 0, and also merge `stopped` flag check into `startOperation()` to simplify the logic. This is not super performance sensitive, so using very tight synchronized block should be fine

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java [hunk 1]
```diff
@@ -80,17 +79,18 @@ public abstract class BasePartitionUpsertMetadataManager implements PartitionUps
...
-  protected final AtomicInteger _numPendingOperations = new AtomicInteger(1);
-
   protected long _lastOutOfOrderEventReportTimeNs = Long.MIN_VALUE;
   protected int _numOutOfOrderEvents = 0;
 
   // Used to maintain the largestSeenComparisonValue to avoid handling out-of-ttl segments/records.
   // If upsertTTL enabled, we will keep track of largestSeenComparisonValue to compute expired segments.
   protected volatile double _largestSeenComparisonValue;
 
+  private boolean _stopped;
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:89]: add a comment that those 3 object states are sync'ed with object's intrinsic lock and condition

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java [hunk 2]
```diff
@@ -702,29 +674,50 @@ public void removeExpiredPrimaryKeys() {
...
   @Override
-  public void stop() {
+  public synchronized void stop() {
+    if (_stopped) {
+      _logger.warn("Metadata manager is already stopped");
+      return;
+    }
     _stopped = true;
-    int numPendingOperations = _numPendingOperations.decrementAndGet();
+    _numPendingOperations--;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:699]: There is a `Preconditions.checkState` in `close()` to ensure `stop()` is already called

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java [hunk 3]
```diff
@@ -702,29 +674,50 @@ public void removeExpiredPrimaryKeys() {
...
+    while (_numPendingOperations != 0) {
+      _logger.info("Waiting for {} pending operations to finish", _numPendingOperations);
+      try {
+        wait();
+      } catch (InterruptedException e) {
+        throw new RuntimeException(
+            String.format("Interrupted while waiting for %d pending operations to finish", _numPendingOperations), e);
       }
     }
     doClose();
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:723]: `synchronized` is reentrant lock so it should never deadlock on itself. We don't want these method to be called in parallel by multiple threads, so `synchronized` should be the right way to achieve that

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:89]: Good point, added

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:724]: ah!

**Review by klsince — APPROVED** [review]

**Pull request merged**